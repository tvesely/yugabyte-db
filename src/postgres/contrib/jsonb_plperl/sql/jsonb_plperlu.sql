CREATE EXTENSION jsonb_plperlu CASCADE;


CREATE FUNCTION testHVToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
$val = {a => 1, b => 'boo', c => undef};
return $val;
$$;

SELECT testHVToJsonb();


CREATE FUNCTION testAVToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
$val = [{a => 1, b => 'boo', c => undef}, {d => 2}];
return $val;
$$;

SELECT testAVToJsonb();


CREATE FUNCTION testSVToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
$val = 1;
return $val;
$$;

SELECT testSVToJsonb();


-- this revealed a bug in the original implementation
CREATE FUNCTION testRegexpResultToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
return ('1' =~ m(0\t2));
$$;

SELECT testRegexpResultToJsonb();


CREATE FUNCTION roundtrip(val jsonb, ref text = '') RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
use Data::Dumper;
$Data::Dumper::Sortkeys = 1;
$Data::Dumper::Indent = 0;
elog(INFO, Dumper($_[0]));
die 'unexpected '.(ref($_[0]) || 'not a').' reference'
    if ref($_[0]) ne $_[1];
return $_[0];
$$;


SELECT roundtrip('null') is null;
SELECT roundtrip('1');
SELECT roundtrip('1E+131071');
SELECT roundtrip('-1');
SELECT roundtrip('1.2');
SELECT roundtrip('-1.2');
SELECT roundtrip('"string"');
SELECT roundtrip('"NaN"');

SELECT roundtrip('true');
SELECT roundtrip('false');

SELECT roundtrip('[]', 'ARRAY');
SELECT roundtrip('[null, null]', 'ARRAY');
SELECT roundtrip('[1, 2, 3]', 'ARRAY');
SELECT roundtrip('[-1, 2, -3]', 'ARRAY');
SELECT roundtrip('[1.2, 2.3, 3.4]', 'ARRAY');
SELECT roundtrip('[-1.2, 2.3, -3.4]', 'ARRAY');
SELECT roundtrip('["string1", "string2"]', 'ARRAY');
SELECT roundtrip('[["string1", "string2"]]', 'ARRAY');

SELECT roundtrip('{}', 'HASH');
SELECT roundtrip('{"1": null}', 'HASH');
SELECT roundtrip('{"1": 1}', 'HASH');
SELECT roundtrip('{"1": -1}', 'HASH');
SELECT roundtrip('{"1": 1.1}', 'HASH');
SELECT roundtrip('{"1": -1.1}', 'HASH');
SELECT roundtrip('{"1": "string1"}', 'HASH');

SELECT roundtrip('{"1": {"2": [3, 4, 5]}, "2": 3}', 'HASH');


\set VERBOSITY terse \\ -- suppress cascade details
DROP EXTENSION plperlu CASCADE;
