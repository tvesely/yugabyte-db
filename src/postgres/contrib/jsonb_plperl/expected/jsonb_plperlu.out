CREATE EXTENSION jsonb_plperlu CASCADE;
NOTICE:  installing required extension "plperlu"
CREATE FUNCTION testHVToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
$val = {a => 1, b => 'boo', c => undef};
return $val;
$$;
SELECT testHVToJsonb();
          testhvtojsonb          
---------------------------------
 {"a": 1, "b": "boo", "c": null}
(1 row)

CREATE FUNCTION testAVToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
$val = [{a => 1, b => 'boo', c => undef}, {d => 2}];
return $val;
$$;
SELECT testAVToJsonb();
                testavtojsonb                
---------------------------------------------
 [{"a": 1, "b": "boo", "c": null}, {"d": 2}]
(1 row)

CREATE FUNCTION testSVToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
$val = 1;
return $val;
$$;
SELECT testSVToJsonb();
 testsvtojsonb 
---------------
 1
(1 row)

-- this revealed a bug in the original implementation
CREATE FUNCTION testRegexpResultToJsonb() RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
return ('1' =~ m(0\t2));
$$;
SELECT testRegexpResultToJsonb();
 testregexpresulttojsonb 
-------------------------
 0
(1 row)

CREATE FUNCTION roundtrip(val jsonb) RETURNS jsonb
LANGUAGE plperlu
TRANSFORM FOR TYPE jsonb
AS $$
return $_[0];
$$;
SELECT roundtrip('null');
 roundtrip 
-----------
 null
(1 row)

SELECT roundtrip('1');
 roundtrip 
-----------
 1
(1 row)

SELECT roundtrip('1E+131071');
ERROR:  cannot convert infinity to jsonb
CONTEXT:  PL/Perl function "roundtrip"
SELECT roundtrip('-1');
 roundtrip 
-----------
 -1
(1 row)

SELECT roundtrip('1.2');
 roundtrip 
-----------
 1.2
(1 row)

SELECT roundtrip('-1.2');
 roundtrip 
-----------
 -1.2
(1 row)

SELECT roundtrip('"string"');
 roundtrip 
-----------
 "string"
(1 row)

SELECT roundtrip('"NaN"');
 roundtrip 
-----------
 "NaN"
(1 row)

SELECT roundtrip('true');
 roundtrip 
-----------
 1
(1 row)

SELECT roundtrip('false');
 roundtrip 
-----------
 0
(1 row)

SELECT roundtrip('[]');
 roundtrip 
-----------
 []
(1 row)

SELECT roundtrip('[null, null]');
  roundtrip   
--------------
 [null, null]
(1 row)

SELECT roundtrip('[1, 2, 3]');
 roundtrip 
-----------
 [1, 2, 3]
(1 row)

SELECT roundtrip('[-1, 2, -3]');
  roundtrip  
-------------
 [-1, 2, -3]
(1 row)

SELECT roundtrip('[1.2, 2.3, 3.4]');
    roundtrip    
-----------------
 [1.2, 2.3, 3.4]
(1 row)

SELECT roundtrip('[-1.2, 2.3, -3.4]');
     roundtrip     
-------------------
 [-1.2, 2.3, -3.4]
(1 row)

SELECT roundtrip('["string1", "string2"]');
       roundtrip        
------------------------
 ["string1", "string2"]
(1 row)

SELECT roundtrip('{}');
 roundtrip 
-----------
 {}
(1 row)

SELECT roundtrip('{"1": null}');
  roundtrip  
-------------
 {"1": null}
(1 row)

SELECT roundtrip('{"1": 1}');
 roundtrip 
-----------
 {"1": 1}
(1 row)

SELECT roundtrip('{"1": -1}');
 roundtrip 
-----------
 {"1": -1}
(1 row)

SELECT roundtrip('{"1": 1.1}');
 roundtrip  
------------
 {"1": 1.1}
(1 row)

SELECT roundtrip('{"1": -1.1}');
  roundtrip  
-------------
 {"1": -1.1}
(1 row)

SELECT roundtrip('{"1": "string1"}');
    roundtrip     
------------------
 {"1": "string1"}
(1 row)

SELECT roundtrip('{"1": {"2": [3, 4, 5]}, "2": 3}');
            roundtrip            
---------------------------------
 {"1": {"2": [3, 4, 5]}, "2": 3}
(1 row)

\set VERBOSITY terse \\ -- suppress cascade details
DROP EXTENSION plperlu CASCADE;
NOTICE:  drop cascades to 6 other objects
